corporate/lib/stripe.py

```diff
@@ -1001,29 +1016,29 @@ def create_stripe_payment_intent(
 
         assert stripe_customer.invoice_settings is not None
         assert stripe_customer.invoice_settings.default_payment_method is not None
+        stripe_invoice = None
         try:
-            # Try to charge user immediately, and if that fails, we inform the user about the failure.
-            stripe_payment_intent = stripe.PaymentIntent.create(
-                amount=payment_intent_data.amount,
-                currency="usd",
-                customer=customer.stripe_customer_id,
-                description=payment_intent_data.description,
-                receipt_email=payment_intent_data.email,
-                confirm=True,
-                statement_descriptor=payment_intent_data.plan_name,
-                metadata=metadata,
-                off_session=True,
-                payment_method=stripe_customer.invoice_settings.default_payment_method,
+            stripe_invoice = self.generate_invoice_for_upgrade(
+                customer,
+                metadata["price_per_license"],
+                metadata["fixed_price"],
+                metadata["licenses"],
+                metadata["plan_tier"],
+                metadata["billing_schedule"],
+                charge_automatically=True,
             )
+            # TODO: Attach metadata to stripe invoice.
+            # Stripe takes its sweet hour to charge customers after creating an invoice.
+            # Since we want to charge customers immediately, we charge them manually.
+            # Then poll for the status of the invoice to see if the payment succeeded.
+            stripe_invoice = stripe.Invoice.pay(stripe_invoice.id)
         except stripe.CardError as e:
+            if stripe_invoice is not None:
+                # Void invoice to avoid double charging if customer tries to upgrade again.
+                stripe.Invoice.void_invoice(stripe_invoice.id)
```

Comment:
Would it perhaps be cleaner to just disable auto-collection of the invoice (https://stripe.com/docs/invoicing/integration/automatic-advancement-collection#update-auto) in this case? (or perhaps to have created this invoice with `auto_advance=False` to begin with, if we end up calling `.pay` explicitly)

I'm just thinking that when we do `.finalize_invoice` earlier, the customer gets an email with the invoice - and it may be a better flow for it to stay valid even if the payment attempt failed; so that they can just go back to the invoice, fix things and pay?

Though I guess doing that would also add the necessity of looking up such older invoices that may have been created if the user **did** end up re-doing the flow, and voiding them all when we receive event about the latest invoice being paid successfully? To not leave old, open invoices laying around

Reply:
Yeah, it definitely make sense to smooth things further here. I will see what's best we can do.